File: JavaScript/WebsiteCode/LeechCode/Medium/MaximumNumberOfIntegersToChooseFromARangeI.js

// 26ms 62.36MB
/**
 * @param {number[]} banned
 * @param {number} n
 * @param {number} maxSum
 * @return {number}
 */
var maxCount = function (banned, n, maxSum) {
    const set = new Set()

    for (const item of banned) {
        if (item <= n) {
            set.add(item)
        }
    }
    let ret = 0;
    let sum = 0;

    for (let i = 1; i <= n; i++) {
        if (set.has(i) === false) {
            sum += i

            if (sum > maxSum)
                break;

            ret++

        }
    }

    return ret
};

// 44ms 65.04MB
/**
 * @param {number[]} banned
 * @param {number} n
 * @param {number} maxSum
 * @return {number}
 */
var maxCount = function(banned, n, maxSum) {
    const bannedSet = new Set(banned); // Use a set for faster lookups
        let sum = 0, count = 0;

        for (let i = 1; i <= n; i++) {
            if (!bannedSet.has(i) && sum + i <= maxSum) { // Skip banned numbers
                sum += i;
                count++;
            }
        }

        return count;
};

File: Python/WebsiteCode/LeetCode/Medium/Array/MaximumNumberOfIntegersToChooseFromARangeI.py

from ast import List

# 52ms 19.13MB
class Solution:
    def maxCount(self, banned: List[int], n: int, maxSum: int) -> int:
        ans = 0
        temp = 0
        banned = set(banned)
        for i in range(1, n + 1):
            if temp >= maxSum or temp + i > maxSum: break
            elif i not in banned:
                temp += i
                ans += 1
        return ans

# 43ms 19.09MB
class Solution:
    def maxCount(self, banned: list[int], n: int, maxSum: int) -> int:
        # Store banned numbers in a set for quick lookup
        banned_set = set(banned)

        count = 0

        # Try each number from 1 to n
        for num in range(1, n + 1):
            # Skip banned numbers
            if num in banned_set:
                continue

            # Return if adding the current number exceeds maxSum
            if maxSum - num < 0:
                return count

            # Include current number
            maxSum -= num
            count += 1

        return count